dart/common/detail/AddonManager.h

```diff
@@ -0,0 +1,177 @@
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_DETAIL_ADDONMANAGER_H_
+#define DART_COMMON_DETAIL_ADDONMANAGER_H_
+
+#include "dart/common/AddonManager.h"
+
+#define DART_COMMON_CHECK_ILLEGAL_ADDON_ERASE( Func, it, ReturnType )\
+  if(it ->second && !it ->second->isOptional(this))\
+  {\
+    dterr << "[AddonManager::" #Func << "] Illegal request to remove Addon!\n";\
+    assert(false);\
+    return ReturnType ;\
+  }
+
+namespace dart {
+namespace common {
+
+//==============================================================================
+template <class T>
+bool AddonManager::has() const
+{
+  return (get<T>() != nullptr);
+}
+
+//==============================================================================
+template <class T>
+T* AddonManager::get()
+{
+  AddonMap::iterator it = mAddonMap.find( typeid(T) );
+  if(mAddonMap.end() == it)
+    return nullptr;
+
+  return static_cast<T*>(it->second.get());
+}
+
+//==============================================================================
+template <class T>
+const T* AddonManager::get() const
+{
+  return const_cast<AddonManager*>(this)->get<T>();
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::set(const T* addon)
+{
+  if(addon)
+  {
+    mAddonMap[typeid(T)] = addon->cloneAddon(this);
+    becomeManager(mAddonMap[typeid(T)].get(), false);
+  }
+  else
+  {
+    mAddonMap[typeid(T)] = nullptr;
+  }
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::set(std::unique_ptr<T>&& addon)
+{
+  mAddonMap[typeid(T)] = std::move(addon);
+  becomeManager(mAddonMap[typeid(T)].get(), true);
+}
+
+//==============================================================================
+template <class T, typename ...Args>
+T* AddonManager::create(Args&&... args)
+{
+  T* addon = new T(this, std::forward<Args>(args)...);
+  mAddonMap[typeid(T)] = std::unique_ptr<T>(addon);
+  becomeManager(addon, false);
+
+  return addon;
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::erase()
+{
+  AddonMap::iterator it = mAddonMap.find( typeid(T) );
+  DART_COMMON_CHECK_ILLEGAL_ADDON_ERASE(erase, it, DART_BLANK)
+  if(mAddonMap.end() != it)
+    it->second = nullptr;
```

Comment:
The two reasons definitely make sense. I realized that the modifiers of managers are not virtual functions. 

:+1: to make a note for this.

Now I'm thinking if it would be a good design that a manager has another fixed size map (or even simpler data structure) for special addons. Then we wouldn't need to worry about the points you mentioned when both of general addons and specialized addons are stored in a single map. By the way, this wouldn't impact the API so (if this makes sense) we can discuss this later.